Determine page of chargeable vehicles

kingarj opened the pull request:
This PR comprises:
-  happy path functionality for fetching only chargeable vehicles for inclusion in the payments matrix.
- checking for entrant payments that have already been made for that particular combination of vrn, date (in the 13 day payment window) and Clean Air Zone.

plenza commented on src/it/java/uk/gov/caz/psr/ExternalCallsIT.java:
```diff
@@ -64,6 +75,44 @@ public void mockVccsComplianceCallError(String vrn, int statusCode) {
+    accountsMockServer
+      .when(requestGet("/v1/accounts/" + accountId + "/vehicles/sorted-page"), exactly(1))
+      .respond(response(responseFile, 200));
+  }
```
I'm wondering if word "Charges" and "Chargeable" has any meaning in Accounts context. Accounts returns list (page) of Fleet vehicles and that's all they do. The "chargeability" aspect is added later by Payment service logic. So maybe we should name these methods more clearly?

plenza commented on src/it/java/uk/gov/caz/psr/RetrieveChargeableAccountVehiclesIT.java:
```diff
@@ -0,0 +1,217 @@
+  }
+  
+  @Test
+  public void shouldReturnHttpOkWhenEntrantPaymentsExist() throws JsonProcessingException {
```
Just to confirm: for now this is only a placeholder test because it will extended with paid days on a later stage?

kingarj replied:
I wasn't calling the right assertion method here - this now checks that the entrant payment (as inserted into the entrant_payment table at the beginning of the test, has been correctly returned and associated with the correct VRN.

plenza commented on src/it/java/uk/gov/caz/psr/RetrieveChargeableAccountVehiclesIT.java:
```diff
@@ -0,0 +1,217 @@
+        ACCOUNT_ID, "ABC123", EMPTY_CURSOR_RESPONSE);
+    List<String> vrns = getAccountVehicleVrnsFromFile(PREVIOUS_CURSOR_RESPONSE);
+    List<String> chargeable = new ArrayList<String>();
+    for (int i = 0; i < vrns.size(); i++) {
```
There is some duplication with previous test: it would be good to keep them as neat as possible so maybe you could extract common setup to private helpers?

plenza commented on src/it/java/uk/gov/caz/psr/journeys/RetrieveChargeableAccountVehiclesJourneyAssertion.java:
```diff
@@ -0,0 +1,103 @@
+  
+  public RetrieveChargeableAccountVehiclesJourneyAssertion 
+  whenRequestIsMadeToRetrieveChargeableAccountVehicles() {
+    RestAssured.basePath = AccountsController.ACCOUNTS_PATH;
```
This should probably be placed in the test setup to be more visible

plenza commented on src/it/java/uk/gov/caz/psr/RetrieveChargeableAccountVehiclesIT.java:
```diff
@@ -0,0 +1,217 @@
+      .whenRequestIsMadeToRetrieveChargeableAccountVehicles()
+      .then()
+      .responseIsReturnedWithHttpOkStatusCode()
+      .responseContainsExpectedData(chargeable, vrns.get(0), null);
```
last vrn is `null`? Is this correct?

plenza commented on src/it/java/uk/gov/caz/psr/journeys/RetrieveChargeableAccountVehiclesJourneyAssertion.java:
```diff
@@ -0,0 +1,103 @@
+
+  public void responseContainsExpectedData(List<String> expectedVrns, 
+      String firstVrn, String lastVrn) {
+    List<PaidPaymentsResult> results = this.responseDto.getPaidPayments().getResults();
```
I don't understand we `getPaidPayments` are being queried here. I thought it is supposed to list all vehicles that need to be paid for not already paid ones?

kingarj replied:
That is poor naming convention on my part - @cruddasj is renaming this to chargeableAccountVehicles in a forthcoming PR.

plenza commented on src/main/java/uk/gov/caz/psr/dto/PaidPaymentsResponse.java:
```diff
@@ -31,6 +31,7 @@ public static PaidPaymentsResponse from(Map<String, List<EntrantPayment>> result
     List<PaidPaymentsResult> mappedResults = results.entrySet()
         .stream()
         .map(entry -> buildPaidPaymentResultFrom(entry.getKey(), entry.getValue()))
+        .sorted((object1, object2) -> object1.getVrn().compareTo(object2.getVrn()))
```
You may want to try `.sorted(Comparator.comparing(PaidPaymentsResponse::getVrn)` for some more readibility

plenza commented on src/main/java/uk/gov/caz/psr/service/AccountService.java:
```diff
@@ -50,6 +59,95 @@ public AccountVehicleRetrievalResponse retrieveAccountVehicles(UUID accountId,
+    String vrnCursor = vrn;
+    
+    while (results.size() < (pageSize + 1) && !lastPage) {
+      // get triple the number of vrns as will be on page to reduce overall request numbers
```
Good idea!

plenza commented on src/main/java/uk/gov/caz/psr/service/AccountService.java:
```diff
@@ -50,6 +59,95 @@ public AccountVehicleRetrievalResponse retrieveAccountVehicles(UUID accountId,
+      List<String> accountVrns = getAccountVrns(accountId, direction, pageSize * 3, vrnCursor);
+      
+      // check if the end of pages has been reached, if not set new cursor
+      if (accountVrns.size() < pageSize) {
```
Should this perhaps be `pageSize * 3`? Not a bug but a small optimization which may cut last 2 calls.

plenza commented on src/main/java/uk/gov/caz/psr/service/AccountService.java:
```diff
@@ -50,6 +59,95 @@ public AccountVehicleRetrievalResponse retrieveAccountVehicles(UUID accountId,
+      
+      List<String> chargeableVrns = getChargeableVrnsFromVcc(accountVrns, cleanAirZoneId);
+      
+      if (! chargeableVrns.isEmpty()) {
```
Please remove space after !

plenza commented on src/main/java/uk/gov/caz/psr/service/AccountService.java:
```diff
@@ -50,6 +59,95 @@ public AccountVehicleRetrievalResponse retrieveAccountVehicles(UUID accountId,
+    Collections.sort(accountVrns);
+    if (direction.equals("next")) {
+      return accountVrns.get(accountVrns.size() - 1);
+    } else if (direction.equals("previous")) {
```
You can remove else keywords due to returns inside ifs

cruddasj commented:
@kingarj - as discussed yesterday, in the event we can't determine the vehicle type (r.g. 404 or 422), the vehicle should not appear in the list from the chargeable vehicles endpoint. Please can you revise the logic and action @plenza's review comments today?

kingarj commented:
@plenza many thanks for your review, please hold fire on re-review as I am just looking at a small improvement to batch asynchronous requests to VCCS.

kingarj commented:
@plenza I've addressed your comments, and added some small enhancements such as the piece around excluding unknown and undeterminable types of vehicles. Are you happy to give this a quick re-review?

cruddasj commented on src/main/java/uk/gov/caz/psr/service/AccountService.java:
```diff
@@ -50,6 +59,118 @@ public AccountVehicleRetrievalResponse retrieveAccountVehicles(UUID accountId,
+      String cleanAirZoneId, int pageSize) {
+    List<String> results = new ArrayList<String>();
+    // split accountVrns into chunks
+    List<List<String>> accountVrnChunks = Lists.partition(accountVrns, 10);
```
Could we pull this currently fixed value from the querystring param received at the controller layer? That way if we needed to change the size of how many vehicles are presented, only a UI layer update would be needed. Not sure if this is the same as the pageSize variable on this method signature?

kingarj replied:
Hmm, my concern with that would be that if you bumped the pageSize var up to say 100 (for instance, if the test team was doing some robust testing of this endpoint), you could end up making that number of asynchronous batch requests to VCCS. What about a variable pulled from application.yml?

cruddasj replied:
I think the challenge we would have is that in the above scenario we would face a problem that the application wouldn't be consistent through between front-end and back-end...

cruddasj replied:
I think the better approach is to pull it from the variable and we make sure the UI can't be manipulated by query string modification
